src/awkward/_connect/numba/layoutbuilder.py

```diff
@@ -0,0 +1,2547 @@
+            numba.from_dtype(val._offsets.dtype), val._content, val._parameters
+        )
+
+    elif isinstance(val, Numpy):
+        return NumpyType(numba.from_dtype(val._data.dtype), val._parameters)
+
+    elif isinstance(val, Record):
+        return RecordType(val._contents, val._fields, val._parameters)
+
+    elif isinstance(val, Regular):
+        return RegularType(val._content, val._size, val._parameters)
+
+    elif isinstance(val, Tuple):
+        return TupleType(val._contents, val._parameters)
+
+    elif isinstance(val, Union):
+        return UnionType(
+            numba.from_dtype(val._tags.dtype),
+            numba.from_dtype(val._index.dtype),
+            val._contents,
+            val._parameters,
+        )
+
+    elif isinstance(val, Unmasked):
+        return UnmaskedType(val._content, val._parameters)
+
+    else:
+        raise TypeError("unrecognized LayoutBuilder type")
+
+
+@numba.extending.overload(len)
+def LayoutBuilderType_len(builder):
+    if isinstance(
+        builder,
+        (
+            BitMaskedType,
+            ByteMaskedType,
+            EmptyType,
+            IndexedOptionType,
+            ListOffsetType,
+            NumpyType,
+            RecordType,
+            RegularType,
+            TupleType,
+            UnionType,
+            UnmaskedType,
+        ),
+    ):
+
+        def len_impl(builder):
+            return builder._length_get()
+
+        return len_impl
+
+
+########## Numpy ############################################################
+
+
+@final
+class Numpy(LayoutBuilder):
+    def __init__(self, dtype, *, parameters=None, initial=1024, resize=8.0):
+        self._data = GrowableBuffer(dtype=dtype, initial=initial, resize=resize)
+        self._parameters = parameters
+
+    @classmethod
+    def _from_buffer(cls, data):
+        out = cls.__new__(cls)
+        out._data = data
+        out._parameters = None
+        return out
+
+    def __repr__(self):
+        return f"ak.numba.lb.Numpy({self._data.dtype}, parameters={self._parameters})"
+
+    def numbatype(self):
+        return NumpyType(
+            numba.from_dtype(self.dtype), numba.types.StringLiteral(self._parameters)
+        )
+
+    @property
+    def _length(self):
+        return len(self._data)
+
+    def __len__(self):
+        return self._length
+
+    @property
+    def dtype(self):
+        return self._data.dtype
+
+    @property
+    def parameters(self):
+        return self._parameters
+
+    def append(self, x):
+        self._data.append(x)
+
+    def extend(self, data):
+        self._data.extend(data)
+
+    def clear(self):
+        self._data.clear()
+
+    def is_valid(self, error: str):
+        return True
+
+    def snapshot(self) -> ArrayLike:
+        """
+        Converts the currently accumulated data into an #ak.Array.
+        """
+        return ak.Array(
+            ak.contents.NumpyArray(self._data.snapshot(), parameters=self._parameters)
+        )
```

Comment:
Since these are layout builders, the `snapshots` should build layouts (`ak.contents.Content` subclasses).

```suggestion
    def snapshot(self) -> ak.contents.Content:
        return ak.contents.NumpyArray(self._data.snapshot(), parameters=self._parameters)
```

In Python, you have the `snapshot` methods returning `ak.Array` (and since you know the exact type, you didn't need to annotate it as `ArrayLike`), and then the recursion had to unpack each level down with `.layout` in order to pack it back up with a new `ak.Array`. Returning the layout directly is a lot more straightforward.

If you want to have a convenience method returning an `ak.Array`, you can have two functions: one recurses, and the other packages up the output of the recursive function in the high-level wrapper. For instance, you could have this:

```python
class LayoutBuilder:  # the abstract class
    def snapshot(self) -> ak.Array:
        return ak.Array(self.snapshot_layout())

class Numpy(LayoutBuilder):  # each subclass
    def snapshot_layout(self) -> ak.contents.Content:
        # recursively calls `snapshot_layout`
```

(I don't have opinions on whether the recursive function is named `snapshot` or `snapshot_layout` and the wrapping function is named `snapshot_array` or `snapshot`, as long as they're a consistent pair.)